teeracle/src/lib.rs

```diff
@@ -51,31 +51,74 @@ pub mod pallet {
 		/// The overarching event type.
 		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
 		type WeightInfo: WeightInfo;
+		/// Max number of whitelisted oracles allowed
+		#[pallet::constant]
+		type MaxOracles: Get<u32>;
 	}
 
+	/// Exchange rates chain's cryptocurrency/currency
 	#[pallet::storage]
 	#[pallet::getter(fn exchange_rate)]
-	pub(super) type ExchangeRates<T> =
+	pub(super) type ExchangeRates<T: Config> =
 		StorageMap<_, Blake2_128Concat, CurrencyString, ExchangeRate, ValueQuery>;
 
+	/// whitelist of trusted exchange rate oracles
+	#[pallet::storage]
+	#[pallet::getter(fn whitelist)]
+	pub(super) type Whitelist<T: Config> =
+		StorageValue<_, WeakBoundedVec<[u8; 32], T::MaxOracles>, ValueQuery>;
+
 	#[pallet::event]
 	#[pallet::generate_deposit(pub(super) fn deposit_event)]
 	pub enum Event<T: Config> {
 		/// The exchange rate of currency was set/updated. \[currency], [new value\]
 		ExchangeRateUpdated(CurrencyString, Option<ExchangeRate>),
 		ExchangeRateDeleted(CurrencyString),
+		AddedToWhitelist([u8; 32]),
+		RemovedFromWhitelist([u8; 32]),
+		OracleWhitelistCleared,
 	}
 
 	#[pallet::error]
 	pub enum Error<T> {
 		InvalidCurrency,
+		/// Too many oracles in the whitelist.
+		TooManyOracles,
```

Comment:
1. You are not whitelisting Oracles (service instances), you are whitelisting Enclaves (binaries). I suggest you refactor `Oracle` to `MrEnclave` here and everywhere else where it makes sense
2. These are all "whitelisting" errors, so I think they should all have "Whitelist" in their name

Reply:
suggestion:
```
MrEnclaveWhitelistOverflow
MrEnclaveNotWhitelisted
MrEnclaveAlreadyWhitelisted
```

Reply:
I think these errors should be user-friendly. And I'm not sure that `MrEnclaveNotWhitelisted`, ... is clear for a dummy user.
From a user's perspective, is it wrong to say it is about the whitelisting of oracles?  In the issue description, you wrote "_the teeracle should maintain a whitelist of oracles_".

Reply:
It seems I wasn't precise. What we are whitelisting is MrEnclave. But it is true that MrEnclave is very SGX specific and not generic for all TEE technologies.
What about `Release`? Each oracle service runs a specific release. And MrEnclave identifies a release unambiguously

Reply:
OK